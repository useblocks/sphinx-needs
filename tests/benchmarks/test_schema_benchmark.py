from collections.abc import Callable
from pathlib import Path
from textwrap import dedent

import pytest
import sphinx
import syrupy
from jinja2 import Template
from sphinx.testing.util import SphinxTestApp
from syrupy.extensions.amber import AmberSnapshotExtension
from syrupy.location import PyTestLocation

# split snapshot dir as warning (sub-)types are emitted only for Sphinx >= 8
SNAPSHOT_DIR = (
    "__snapshots__sphinx_lt_8"
    if sphinx.version_info[0] < 8
    else "__snapshots__sphinx_ge_8"
)
"""Snapshot directory name."""


class DifferentDirectoryExtension(AmberSnapshotExtension):
    """Overwrite the directory name for the snapshot files."""

    @classmethod
    def dirname(cls, *, test_location: "PyTestLocation") -> str:
        return str(Path(test_location.filepath).parent.joinpath(SNAPSHOT_DIR))


@pytest.fixture
def snapshot(snapshot: syrupy.SnapshotAssertion):
    return snapshot.use_extension(DifferentDirectoryExtension)


@pytest.mark.parametrize(
    "need_cnt",
    [
        10,
        100,
        1_000,
        # 5_000,
        # 10_000,
    ],
)
@pytest.mark.benchmark
def test_schema_benchmark(
    tmpdir: Path,
    make_app: Callable[[], SphinxTestApp],
    need_cnt: int,
    snapshot: syrupy.SnapshotAssertion,
    get_warnings_list: Callable[[SphinxTestApp], list[str]],
):
    """
    Benchmark on many needs with thousands of warnings.

    This also tests the dedicated schema builder.
    """
    assert need_cnt % 10 == 0, "need_cnt must be a multiple of 10"
    page_cnt = int(need_cnt / 10)

    this_file_dir = Path(__file__).parent

    src_dir = this_file_dir / ".." / "doc_test" / "doc_schema_benchmark"
    page_template_path = src_dir / "page.rst.j2"
    with page_template_path.open() as fp:
        template_content = fp.read()

    template = Template(template_content)
    pages_dir = Path(tmpdir) / "pages"
    pages_dir.mkdir(exist_ok=True)
    toctree_content = dedent(
        """
        .. toctree::
           :maxdepth: 2

        """
    )

    width = len(str(page_cnt))
    for i in range(1, page_cnt + 1):
        i_fmt = f"{i:0{width}d}"
        page_rst_content = template.render(page_nr=i_fmt)

        page_name = f"page_{i_fmt}"
        page_file = f"{page_name}.rst"
        page_rst_path = pages_dir / page_file
        page_rst_path.write_text(page_rst_content, encoding="utf-8")
        toctree_content += f"   pages/{page_name}\n"

    index_file = tmpdir / "index.rst"
    index_file.write_text(toctree_content, encoding="utf-8")

    copy_files = [
        src_dir / "conf.py",
        src_dir / "schemas.json",
        src_dir / "ubproject.toml",
    ]
    for copy_file in copy_files:
        dst_file = tmpdir / copy_file.name
        dst_file.write_text(copy_file.read_text(), encoding="utf-8")

    app: SphinxTestApp = make_app(
        # the schema builder does only validate, no output
        buildername="schema",
        srcdir=Path(tmpdir),
        freshenv=True,
    )
    app.build()

    assert app.statuscode == 0

    warnings = get_warnings_list(app)
    expected_warnings_per_page = 7
    assert len(warnings) == expected_warnings_per_page * page_cnt

    if need_cnt < 500:
        # keep snapshot small;
        assert warnings == snapshot

    app.cleanup()
