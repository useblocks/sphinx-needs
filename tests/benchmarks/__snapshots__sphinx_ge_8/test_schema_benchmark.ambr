# serializer version: 1
# name: test_schema_benchmark[100]
  list([
    '''
      Need 'FEAt_P01' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      FEAt_P01
        Schema path:    [0] > local > properties > id > pattern
        User message:   id must be uppercase
        Schema message: 'FEAt_P01' does not match '^[A-Z0-9_]+$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_MISSING_APPROVAL_P01' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_MISSING_APPROVAL_P01
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_MISSING_APPROVAL_P01' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_MISSING_APPROVAL_P01' has validation errors:
        Severity:       violation
        Field:          
        Need path:      SPEC_MISSING_APPROVAL_P01
        Schema path:    spec-approved[2] > local > required
        User message:   Approval required due to high efforts
        Schema message: 'approved' is a required property [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_P01' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_P01
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_P01' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_SAFE_UNSAFE_FEAT_P01' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_SAFE_UNSAFE_FEAT_P01
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_SAFE_UNSAFE_FEAT_P01' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_SAFE_P01' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_SAFE_P01
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_SAFE_P01' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'FEAT_P01' has validation errors:
        Severity:       violation
        Field:          asil
        Need path:      IMPL_SAFE_P01 > links > SPEC_SAFE_UNSAFE_FEAT_P01 > links > FEAT_P01
        Schema path:    safe-impl-[links]->safe-spec-[links]->safe-feat[4] > links > links > local > allOf > 0 > allOf > 0 > properties > asil > enum
        Schema message: 'QM' is not one of ['A', 'B', 'C', 'D'] [sn_schema.network_local_fail]
  
    ''',
    '''
      Need 'FEAt_P02' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      FEAt_P02
        Schema path:    [0] > local > properties > id > pattern
        User message:   id must be uppercase
        Schema message: 'FEAt_P02' does not match '^[A-Z0-9_]+$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_MISSING_APPROVAL_P02' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_MISSING_APPROVAL_P02
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_MISSING_APPROVAL_P02' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_MISSING_APPROVAL_P02' has validation errors:
        Severity:       violation
        Field:          
        Need path:      SPEC_MISSING_APPROVAL_P02
        Schema path:    spec-approved[2] > local > required
        User message:   Approval required due to high efforts
        Schema message: 'approved' is a required property [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_P02' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_P02
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_P02' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_SAFE_UNSAFE_FEAT_P02' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_SAFE_UNSAFE_FEAT_P02
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_SAFE_UNSAFE_FEAT_P02' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_SAFE_P02' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_SAFE_P02
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_SAFE_P02' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'FEAT_P02' has validation errors:
        Severity:       violation
        Field:          asil
        Need path:      IMPL_SAFE_P02 > links > SPEC_SAFE_UNSAFE_FEAT_P02 > links > FEAT_P02
        Schema path:    safe-impl-[links]->safe-spec-[links]->safe-feat[4] > links > links > local > allOf > 0 > allOf > 0 > properties > asil > enum
        Schema message: 'QM' is not one of ['A', 'B', 'C', 'D'] [sn_schema.network_local_fail]
  
    ''',
    '''
      Need 'FEAt_P03' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      FEAt_P03
        Schema path:    [0] > local > properties > id > pattern
        User message:   id must be uppercase
        Schema message: 'FEAt_P03' does not match '^[A-Z0-9_]+$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_MISSING_APPROVAL_P03' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_MISSING_APPROVAL_P03
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_MISSING_APPROVAL_P03' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_MISSING_APPROVAL_P03' has validation errors:
        Severity:       violation
        Field:          
        Need path:      SPEC_MISSING_APPROVAL_P03
        Schema path:    spec-approved[2] > local > required
        User message:   Approval required due to high efforts
        Schema message: 'approved' is a required property [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_P03' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_P03
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_P03' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_SAFE_UNSAFE_FEAT_P03' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_SAFE_UNSAFE_FEAT_P03
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_SAFE_UNSAFE_FEAT_P03' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_SAFE_P03' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_SAFE_P03
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_SAFE_P03' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'FEAT_P03' has validation errors:
        Severity:       violation
        Field:          asil
        Need path:      IMPL_SAFE_P03 > links > SPEC_SAFE_UNSAFE_FEAT_P03 > links > FEAT_P03
        Schema path:    safe-impl-[links]->safe-spec-[links]->safe-feat[4] > links > links > local > allOf > 0 > allOf > 0 > properties > asil > enum
        Schema message: 'QM' is not one of ['A', 'B', 'C', 'D'] [sn_schema.network_local_fail]
  
    ''',
    '''
      Need 'FEAt_P04' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      FEAt_P04
        Schema path:    [0] > local > properties > id > pattern
        User message:   id must be uppercase
        Schema message: 'FEAt_P04' does not match '^[A-Z0-9_]+$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_MISSING_APPROVAL_P04' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_MISSING_APPROVAL_P04
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_MISSING_APPROVAL_P04' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_MISSING_APPROVAL_P04' has validation errors:
        Severity:       violation
        Field:          
        Need path:      SPEC_MISSING_APPROVAL_P04
        Schema path:    spec-approved[2] > local > required
        User message:   Approval required due to high efforts
        Schema message: 'approved' is a required property [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_P04' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_P04
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_P04' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_SAFE_UNSAFE_FEAT_P04' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_SAFE_UNSAFE_FEAT_P04
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_SAFE_UNSAFE_FEAT_P04' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_SAFE_P04' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_SAFE_P04
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_SAFE_P04' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'FEAT_P04' has validation errors:
        Severity:       violation
        Field:          asil
        Need path:      IMPL_SAFE_P04 > links > SPEC_SAFE_UNSAFE_FEAT_P04 > links > FEAT_P04
        Schema path:    safe-impl-[links]->safe-spec-[links]->safe-feat[4] > links > links > local > allOf > 0 > allOf > 0 > properties > asil > enum
        Schema message: 'QM' is not one of ['A', 'B', 'C', 'D'] [sn_schema.network_local_fail]
  
    ''',
    '''
      Need 'FEAt_P05' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      FEAt_P05
        Schema path:    [0] > local > properties > id > pattern
        User message:   id must be uppercase
        Schema message: 'FEAt_P05' does not match '^[A-Z0-9_]+$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_MISSING_APPROVAL_P05' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_MISSING_APPROVAL_P05
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_MISSING_APPROVAL_P05' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_MISSING_APPROVAL_P05' has validation errors:
        Severity:       violation
        Field:          
        Need path:      SPEC_MISSING_APPROVAL_P05
        Schema path:    spec-approved[2] > local > required
        User message:   Approval required due to high efforts
        Schema message: 'approved' is a required property [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_P05' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_P05
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_P05' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_SAFE_UNSAFE_FEAT_P05' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_SAFE_UNSAFE_FEAT_P05
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_SAFE_UNSAFE_FEAT_P05' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_SAFE_P05' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_SAFE_P05
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_SAFE_P05' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'FEAT_P05' has validation errors:
        Severity:       violation
        Field:          asil
        Need path:      IMPL_SAFE_P05 > links > SPEC_SAFE_UNSAFE_FEAT_P05 > links > FEAT_P05
        Schema path:    safe-impl-[links]->safe-spec-[links]->safe-feat[4] > links > links > local > allOf > 0 > allOf > 0 > properties > asil > enum
        Schema message: 'QM' is not one of ['A', 'B', 'C', 'D'] [sn_schema.network_local_fail]
  
    ''',
    '''
      Need 'FEAt_P06' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      FEAt_P06
        Schema path:    [0] > local > properties > id > pattern
        User message:   id must be uppercase
        Schema message: 'FEAt_P06' does not match '^[A-Z0-9_]+$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_MISSING_APPROVAL_P06' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_MISSING_APPROVAL_P06
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_MISSING_APPROVAL_P06' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_MISSING_APPROVAL_P06' has validation errors:
        Severity:       violation
        Field:          
        Need path:      SPEC_MISSING_APPROVAL_P06
        Schema path:    spec-approved[2] > local > required
        User message:   Approval required due to high efforts
        Schema message: 'approved' is a required property [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_P06' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_P06
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_P06' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_SAFE_UNSAFE_FEAT_P06' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_SAFE_UNSAFE_FEAT_P06
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_SAFE_UNSAFE_FEAT_P06' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_SAFE_P06' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_SAFE_P06
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_SAFE_P06' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'FEAT_P06' has validation errors:
        Severity:       violation
        Field:          asil
        Need path:      IMPL_SAFE_P06 > links > SPEC_SAFE_UNSAFE_FEAT_P06 > links > FEAT_P06
        Schema path:    safe-impl-[links]->safe-spec-[links]->safe-feat[4] > links > links > local > allOf > 0 > allOf > 0 > properties > asil > enum
        Schema message: 'QM' is not one of ['A', 'B', 'C', 'D'] [sn_schema.network_local_fail]
  
    ''',
    '''
      Need 'FEAt_P07' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      FEAt_P07
        Schema path:    [0] > local > properties > id > pattern
        User message:   id must be uppercase
        Schema message: 'FEAt_P07' does not match '^[A-Z0-9_]+$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_MISSING_APPROVAL_P07' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_MISSING_APPROVAL_P07
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_MISSING_APPROVAL_P07' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_MISSING_APPROVAL_P07' has validation errors:
        Severity:       violation
        Field:          
        Need path:      SPEC_MISSING_APPROVAL_P07
        Schema path:    spec-approved[2] > local > required
        User message:   Approval required due to high efforts
        Schema message: 'approved' is a required property [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_P07' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_P07
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_P07' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_SAFE_UNSAFE_FEAT_P07' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_SAFE_UNSAFE_FEAT_P07
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_SAFE_UNSAFE_FEAT_P07' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_SAFE_P07' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_SAFE_P07
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_SAFE_P07' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'FEAT_P07' has validation errors:
        Severity:       violation
        Field:          asil
        Need path:      IMPL_SAFE_P07 > links > SPEC_SAFE_UNSAFE_FEAT_P07 > links > FEAT_P07
        Schema path:    safe-impl-[links]->safe-spec-[links]->safe-feat[4] > links > links > local > allOf > 0 > allOf > 0 > properties > asil > enum
        Schema message: 'QM' is not one of ['A', 'B', 'C', 'D'] [sn_schema.network_local_fail]
  
    ''',
    '''
      Need 'FEAt_P08' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      FEAt_P08
        Schema path:    [0] > local > properties > id > pattern
        User message:   id must be uppercase
        Schema message: 'FEAt_P08' does not match '^[A-Z0-9_]+$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_MISSING_APPROVAL_P08' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_MISSING_APPROVAL_P08
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_MISSING_APPROVAL_P08' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_MISSING_APPROVAL_P08' has validation errors:
        Severity:       violation
        Field:          
        Need path:      SPEC_MISSING_APPROVAL_P08
        Schema path:    spec-approved[2] > local > required
        User message:   Approval required due to high efforts
        Schema message: 'approved' is a required property [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_P08' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_P08
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_P08' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_SAFE_UNSAFE_FEAT_P08' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_SAFE_UNSAFE_FEAT_P08
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_SAFE_UNSAFE_FEAT_P08' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_SAFE_P08' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_SAFE_P08
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_SAFE_P08' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'FEAT_P08' has validation errors:
        Severity:       violation
        Field:          asil
        Need path:      IMPL_SAFE_P08 > links > SPEC_SAFE_UNSAFE_FEAT_P08 > links > FEAT_P08
        Schema path:    safe-impl-[links]->safe-spec-[links]->safe-feat[4] > links > links > local > allOf > 0 > allOf > 0 > properties > asil > enum
        Schema message: 'QM' is not one of ['A', 'B', 'C', 'D'] [sn_schema.network_local_fail]
  
    ''',
    '''
      Need 'FEAt_P09' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      FEAt_P09
        Schema path:    [0] > local > properties > id > pattern
        User message:   id must be uppercase
        Schema message: 'FEAt_P09' does not match '^[A-Z0-9_]+$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_MISSING_APPROVAL_P09' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_MISSING_APPROVAL_P09
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_MISSING_APPROVAL_P09' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_MISSING_APPROVAL_P09' has validation errors:
        Severity:       violation
        Field:          
        Need path:      SPEC_MISSING_APPROVAL_P09
        Schema path:    spec-approved[2] > local > required
        User message:   Approval required due to high efforts
        Schema message: 'approved' is a required property [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_P09' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_P09
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_P09' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_SAFE_UNSAFE_FEAT_P09' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_SAFE_UNSAFE_FEAT_P09
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_SAFE_UNSAFE_FEAT_P09' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_SAFE_P09' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_SAFE_P09
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_SAFE_P09' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'FEAT_P09' has validation errors:
        Severity:       violation
        Field:          asil
        Need path:      IMPL_SAFE_P09 > links > SPEC_SAFE_UNSAFE_FEAT_P09 > links > FEAT_P09
        Schema path:    safe-impl-[links]->safe-spec-[links]->safe-feat[4] > links > links > local > allOf > 0 > allOf > 0 > properties > asil > enum
        Schema message: 'QM' is not one of ['A', 'B', 'C', 'D'] [sn_schema.network_local_fail]
  
    ''',
    '''
      Need 'FEAt_P10' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      FEAt_P10
        Schema path:    [0] > local > properties > id > pattern
        User message:   id must be uppercase
        Schema message: 'FEAt_P10' does not match '^[A-Z0-9_]+$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_MISSING_APPROVAL_P10' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_MISSING_APPROVAL_P10
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_MISSING_APPROVAL_P10' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_MISSING_APPROVAL_P10' has validation errors:
        Severity:       violation
        Field:          
        Need path:      SPEC_MISSING_APPROVAL_P10
        Schema path:    spec-approved[2] > local > required
        User message:   Approval required due to high efforts
        Schema message: 'approved' is a required property [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_P10' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_P10
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_P10' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_SAFE_UNSAFE_FEAT_P10' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_SAFE_UNSAFE_FEAT_P10
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_SAFE_UNSAFE_FEAT_P10' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_SAFE_P10' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_SAFE_P10
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_SAFE_P10' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'FEAT_P10' has validation errors:
        Severity:       violation
        Field:          asil
        Need path:      IMPL_SAFE_P10 > links > SPEC_SAFE_UNSAFE_FEAT_P10 > links > FEAT_P10
        Schema path:    safe-impl-[links]->safe-spec-[links]->safe-feat[4] > links > links > local > allOf > 0 > allOf > 0 > properties > asil > enum
        Schema message: 'QM' is not one of ['A', 'B', 'C', 'D'] [sn_schema.network_local_fail]
  
    ''',
  ])
# ---
# name: test_schema_benchmark[10]
  list([
    '''
      Need 'FEAt_P1' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      FEAt_P1
        Schema path:    [0] > local > properties > id > pattern
        User message:   id must be uppercase
        Schema message: 'FEAt_P1' does not match '^[A-Z0-9_]+$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_MISSING_APPROVAL_P1' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_MISSING_APPROVAL_P1
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_MISSING_APPROVAL_P1' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_MISSING_APPROVAL_P1' has validation errors:
        Severity:       violation
        Field:          
        Need path:      SPEC_MISSING_APPROVAL_P1
        Schema path:    spec-approved[2] > local > required
        User message:   Approval required due to high efforts
        Schema message: 'approved' is a required property [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_P1' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_P1
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_P1' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_SAFE_UNSAFE_FEAT_P1' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_SAFE_UNSAFE_FEAT_P1
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_SAFE_UNSAFE_FEAT_P1' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'SPEC_SAFE_P1' has validation errors:
        Severity:       violation
        Field:          id
        Need path:      SPEC_SAFE_P1
        Schema path:    spec[1] > local > properties > id > pattern
        Schema message: 'SPEC_SAFE_P1' does not match '^REQ[a-zA-Z0-9_-]*$' [sn_schema.local_fail]
  
    ''',
    '''
      Need 'FEAT_P1' has validation errors:
        Severity:       violation
        Field:          asil
        Need path:      IMPL_SAFE_P1 > links > SPEC_SAFE_UNSAFE_FEAT_P1 > links > FEAT_P1
        Schema path:    safe-impl-[links]->safe-spec-[links]->safe-feat[4] > links > links > local > allOf > 0 > allOf > 0 > properties > asil > enum
        Schema message: 'QM' is not one of ['A', 'B', 'C', 'D'] [sn_schema.network_local_fail]
  
    ''',
  ])
# ---
